components/eamxx/src/physics/nudging/atmosphere_nudging.cpp

```diff
@@ -0,0 +1,275 @@
+      NudgingData_aft.time = time_step_file*(time_index+1);
+    }
+  
+}
+
+  
+// =========================================================================================
+void Nudging::run_impl (const int dt)
+{
+  using namespace scream::vinterp;
+
+  //Have to add dt because first time iteration is at 0 seconds where you will not have
+  //any data from the field. The timestamp is only iterated at the end of the
+  //full step in scream.
+  auto ts = timestamp()+dt;
+  auto time_since_zero = ts.seconds_from(ts0);
+
+  //Don't nudge for times less than first time step in nudged file to avoid
+  //extrapolation
+  if (time_since_zero < time_step_file){ return;}
+
+  //update time state information and check whether need to update data
+  update_time_step(time_since_zero);
+  
+  //perform time interpolation
+  time_interpolation(time_since_zero);
+
+  auto T_mid       = get_field_out("T_mid").get_view<mPack**>();
+  auto qv          = get_field_out("qv").get_view<mPack**>();
+  auto hw          = get_field_out("horiz_winds").get_view<mPack***>();
+
+  const auto& p_mid    = get_field_in("p_mid").get_view<const mPack**>();
+  auto p_mid_v = view_Nd<mPack,2>("",p_mid.extent_int(0),p_mid.extent_int(1));
+  Kokkos::deep_copy(p_mid_v,p_mid);
+  
+  const view_Nd<mPack,2> T_mid_ext_p(reinterpret_cast<mPack*>(T_mid_ext.data()),
+				     m_num_cols,m_num_src_levs); 
+  const view_Nd<mPack,2> p_mid_ext_p(reinterpret_cast<mPack*>(p_mid_ext.data()),
+				     m_num_cols,m_num_src_levs);
+  const view_Nd<mPack,2> qv_ext_p(reinterpret_cast<mPack*>(qv_ext.data()),
+				  m_num_cols,m_num_src_levs);
+  const view_Nd<mPack,3> hw_ext_p(reinterpret_cast<mPack*>(horiz_winds_ext.data()),
+				  m_num_cols,2,m_num_src_levs);
+
+  //Now perform the vertical interpolation from the external file to the internal
+  //field levels
+  perform_vertical_interpolation<Real,1,2>(p_mid_ext_p,
+                                           p_mid_v,
+                                           T_mid_ext_p,
+                                           T_mid,
+                                           m_num_src_levs,
+                                           m_num_levs);
+
+  perform_vertical_interpolation<Real,1,2>(p_mid_ext_p,
+                                           p_mid_v,
+                                           qv_ext_p,
+                                           qv,
+                                           m_num_src_levs,
+                                           m_num_levs);
+
+  perform_vertical_interpolation<Real,1,3>(p_mid_ext_p,
+                                           p_mid_v,
+                                           hw_ext_p,
+                                           hw,
+                                           m_num_src_levs,
+                                           m_num_levs);
+
+  //This code removes any masked values and sets them to the corresponding
+  //values at the lowest/highest pressure levels from the external file
```

Comment:
Wait, I thought we were using ekat's linear interpolation in such a way that we were not extrapolating. That is, if you pad x1 and y1 passed to lin interp as (pseudo code)
```
L = some-very-large-x (for pressure, it could be 2.0)
x1 = [x1 L]
y1 = [y1 y1(end)]
```
then you will be guaranteed that, ix x2 is larger than the max of the original (pre-padding) values of x1, then y2 will be set to the last entry of y1.

Reply:
Yes, ekat's linear interpolation is not extrapolating, so you could pre-pad and maybe I should have done this. I have to say it seemed easier to remove the masked values after, but maybe it is better to pre-pad. Just so I understand this would entail taking for instance T_mid_ext_p (see below) and adding two more levels, one at the bottom and one at the top and then setting the bottom value to the bottom+1 value and the top to the top-1 value
  const view_Nd<mPack,2> T_mid_ext_p(reinterpret_cast<mPack*>(T_mid_ext.data()),
                                     m_num_cols,m_num_src_levs);
And then as you say for the pressure add two levels with one very large and one very small so that there are never any layers outside those bounds. I guess what was making it difficult was having to define a new T_mid field with two extra levels. Is there an easy way to add levels to a view as opposed to creating a new view with +2 levels and looping over this new one and filling in values from the original (which doesn't seem any easier than looping over to remove masked values)?

Reply:
Will keep as is after conversations via Slack